instrumentation/kafka-clients/src/main/java/brave/kafka/clients/TracingConsumer.java

```diff
@@ -43,38 +43,48 @@
-      handleConsumed(record);
+    if (records.isEmpty()) return records;
+    for (TopicPartition partition : records.partitions()) {
+      List<ConsumerRecord<K, V>> recordsInPartition = records.records(partition);
+      Span consumerSpanForTopicPartition = null;
+      for (int i = 0, length = recordsInPartition.size(); i < length; i++) {
+        ConsumerRecord<K, V> record = recordsInPartition.get(i);
+        TraceContextOrSamplingFlags extracted = extractor.extract(record.headers());
+
+        // If we extracted neither a trace context, nor request-scoped data (extra),
+        // make or reuse a span for this topic/partition
+        if (extracted.samplingFlags() != null && extracted.extra().isEmpty()) {
+          if (consumerSpanForTopicPartition == null) {
+            consumerSpanForTopicPartition = tracing.tracer().nextSpan(extracted);
+            addTagsAndStart(consumerSpanForTopicPartition, partition);
+          }
+          Span span = consumerSpanForTopicPartition; // span is topic/partition scope
+          // no need to remove propagation headers as we failed to extract anything
+          injector.inject(span.context(), record.headers());
+        } else { // we extracted request-scoped data, so cannot share a consumer span.
+          Span span = tracing.tracer().nextSpan(extracted);
+          addTagsAndStart(span, partition);
+          span.finish();  // span won't be shared by other records
+          // remove prior propagation headers from the record
+          tracing.propagation().keys().forEach(key -> record.headers().remove(key));
+          injector.inject(span.context(), record.headers());
+        }
+      }
+      if (consumerSpanForTopicPartition != null) consumerSpanForTopicPartition.finish();
     }
     return records;
   }
 
-  void handleConsumed(ConsumerRecord record) {
-    Span span = startAndFinishConsumerSpan(record);
-    // remove propagation headers from the record
-    tracing.propagation().keys().forEach(key -> record.headers().remove(key));
-
-    // Inject new propagation headers
-    injector.inject(span.context(), record.headers());
-  }
-
-  /**
-   * Start a consumer span child of the producer span. And immediately finish It.
-   */
-  Span startAndFinishConsumerSpan(ConsumerRecord record) {
-    TraceContextOrSamplingFlags extracted = extractor.extract(record.headers());
-    Span span = tracing.tracer().nextSpan(extracted);
-    if (!span.isNoop()) {
-      if (record.key() != null) {
-        span.tag(KafkaTags.KAFKA_KEY_TAG, record.key().toString());
-      }
-      span.tag(KafkaTags.KAFKA_TOPIC_TAG, record.topic());
+  static void addTagsAndStart(Span consumerSpanForTopicPartition, TopicPartition partition) {
+    if (!consumerSpanForTopicPartition.isNoop()) {
+      consumerSpanForTopicPartition
+          .kind(Span.Kind.CONSUMER)
+          .tag(KafkaTags.KAFKA_TOPIC_TAG, partition.topic())
+          .tag(KafkaTags.KAFKA_PARTITION_TAG, String.valueOf(partition.partition()))
```

Comment:
No more KEY_TAG ?
The partition is just an ID, It could be useful but the key still makes sense IMHO.